backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/tournament/domain/Tournament.java

```diff
@@ -0,0 +1,241 @@
+    private Integer quizId;
+
+    @Column(name = "privateTournament")
+    private boolean privateTournament;
+
+    @Column(name = "password")
+    private String password;
+
+    public Tournament() {
+    }
+
+    public Tournament(User user, CourseExecution courseExecution, Set<Topic> topics, TournamentDto tournamentDto) {
+        setStartTime(DateHandler.toLocalDateTime(tournamentDto.getStartTime()));
+        setEndTime(DateHandler.toLocalDateTime(tournamentDto.getEndTime()));
+        setNumberOfQuestions(tournamentDto.getNumberOfQuestions());
+        this.isCanceled = tournamentDto.isCanceled();
+        this.creator = user;
+        setCourseExecution(courseExecution);
+        setTopics(topics);
+        setPassword(tournamentDto.getPassword());
+        setPrivateTournament(tournamentDto.isPrivateTournament());
+    }
+
+    public Integer getId() { return id; }
+
+    public LocalDateTime getStartTime() { return startTime; }
+
+    public void setStartTime(LocalDateTime startTime) {
+        // Added 1 minute as a buffer to take latency into consideration
+        if (startTime == null || (this.endTime != null && this.endTime.isBefore(startTime) ||
+                startTime.plusMinutes(1).isBefore(DateHandler.now()))) {
+            throw new TutorException(TOURNAMENT_NOT_CONSISTENT, "startTime");
+        }
+
+        this.startTime = startTime;
+    }
+
+    public LocalDateTime getEndTime() { return endTime; }
+
+    public void setEndTime(LocalDateTime endTime) {
+        if (endTime == null || (this.startTime != null && endTime.isBefore(this.startTime))) {
+            throw new TutorException(TOURNAMENT_NOT_CONSISTENT, "endTime");
+        }
+
+        this.endTime = endTime;
+    }
+
+    public void setNumberOfQuestions(Integer numberOfQuestions) {
+        if (numberOfQuestions <= 0) {
+            throw new TutorException(TOURNAMENT_NOT_CONSISTENT, "number of questions");
+        }
+        this.numberOfQuestions = numberOfQuestions;
+    }
+
+    public Integer getNumberOfQuestions() { return numberOfQuestions; }
+
+    public User getCreator() { return creator; }
+
+    public boolean isCanceled() { return isCanceled; }
+
+    public void cancel(Integer numberOfAnswers) {
+        this.checkCanChange(numberOfAnswers);
+        this.isCanceled = true;
+    }
+
+    public Set<User> getParticipants() { return participants; }
+
+    public void setCourseExecution(CourseExecution courseExecution) { this.courseExecution = courseExecution; }
+
+    public CourseExecution getCourseExecution() { return courseExecution; }
+
+    public Integer getQuizId() { return quizId; }
+
+    public void setQuizId(Integer quizId) { this.quizId = quizId; }
+
+    public Set<Topic> getTopics() { return topics; }
+
+    public void setTopics(Set<Topic> topics) {
+        for (Topic topic: topics) {
+            checkTopicCourse(topic);
+        }
+
+        this.topics = topics;
+    }
+
+    public void updateTopics(Set<Topic> newTopics) {
```

Comment:
not sure if it would be more efficient to remove all topic and add this set

Reply:
Done